src/vs/workbench/parts/debug/electron-browser/debugInlineDecorators.ts

```diff
@@ -0,0 +1,167 @@
+import { StandardTokenType } from 'vs/editor/common/core/lineTokens';
+
+export const MAX_INLINE_VALUE_LENGTH = 50; // Max string length of each inline 'x = y' string. If exceeded ... is added
+export const MAX_INLINE_DECORATOR_LENGTH = 150; // Max string length of each inline decorator when debugging. If exceeded ... is added
+export const MAX_NUM_INLINE_VALUES = 100; // JS Global scope can have 700+ entries. We want to limit ourselves for perf reasons
+export const MAX_TOKENIZATION_LINE_LEN = 500; // If line is too long, then inline values for the line are skipped
+export const ELLIPSES = '…';
+// LanguageConfigurationRegistry.getWordDefinition() return regexes that allow spaces and punctuation characters for languages like python
+// Using that approach is not viable so we are using a simple regex to look for word tokens.
+export const WORD_REGEXP = /[\$\_A-Za-z][\$\_A-Za-z0-9]*/g;
+
+export function getNameValueMapFromScopeChildren(scopeExpressions: IExpression[][]): IStringDictionary<string> {
+	const nameValueMap = createStringDictionary<string>();
+	let valueCount = 0;
+
+	for (let expressions of scopeExpressions) {
+		for (let expr of expressions) {
+			// Put ellipses in value if its too long. Preserve last char e.g "longstr…" or {a:true, b:true, …}
+			let value = expr.value;
+			if (value && value.length > MAX_INLINE_VALUE_LENGTH) {
+				value = value.substr(0, MAX_INLINE_VALUE_LENGTH - ELLIPSES.length) + ELLIPSES + value[value.length - 1];
+			}
+
+			nameValueMap[expr.name] = value;
+
+			// Limit the size of map. Too large can have a perf impact
+			if (++valueCount >= MAX_NUM_INLINE_VALUES) {
+				return nameValueMap;
+			}
+		}
+	}
+
+	return nameValueMap;
+}
+
+export function getDecorators(nameValueMap: IStringDictionary<string>, wordRangeMap: IStringDictionary<IRange[]>, linesContent: string[]): IDecorationOptions[] {
+	const linesNames = createStringDictionary<IStringDictionary<boolean>>();
+	const names = Object.keys(nameValueMap);
+	const decorators: IDecorationOptions[] = [];
+
+	// Compute unique set of names on each line
+	for (let name of names) {
+		const ranges = wordRangeMap[name];
+		if (ranges) {
+			for (let range of ranges) {
+				const lineNum = range.startLineNumber;
+				if (!linesNames[lineNum]) {
+					linesNames[lineNum] = createStringDictionary<boolean>();
+				}
+				linesNames[lineNum][name] = true;
+			}
+		}
+	}
+
+	// Compute decorators for each line
+	const lineNums = Object.keys(linesNames);
+	for (let lineNum of lineNums) {
+		const uniqueNames = Object.keys(linesNames[lineNum]);
+		const decorator = getDecoratorFromNames(parseInt(lineNum), uniqueNames, nameValueMap, linesContent);
+		decorators.push(decorator);
+	}
+
+	return decorators;
+}
+
+export function getDecoratorFromNames(lineNumber: number, names: string[], nameValueMap: IStringDictionary<string>, linesContent: string[]): IDecorationOptions {
+	const margin = '10px';
+	const backgroundColor = 'rgba(255,200,0,0.2)';
```

Comment:
We would like to change the underlying implementation since it is already causing performance issues for folding. Though I will have to sync with @alexandrudima on this